fish-rust/src/builtins/functions.rs

```diff
@@ -0,0 +1,420 @@
+use super::prelude::*;
+use crate::common::escape_string;
+use crate::common::reformat_for_screen;
+use crate::common::valid_func_name;
+use crate::common::{EscapeFlags, EscapeStringStyle};
+use crate::event::{self};
+use crate::ffi::colorize_shell;
+use crate::function;
+use crate::parser_keywords::parser_keywords_is_reserved;
+use crate::termsize::termsize_last;
+
+struct FunctionsCmdOpts {
+    print_help: bool,
+    erase: bool,
+    list: bool,
+    show_hidden: bool,
+    query: bool,
+    copy: bool,
+    report_metadata: bool,
+    no_metadata: bool,
+    verbose: bool,
+    handlers: bool,
+    handlers_type: WString,
+    description: WString,
+}
+
+impl Default for FunctionsCmdOpts {
+    fn default() -> Self {
+        Self {
+            print_help: false,
+            erase: false,
+            list: false,
+            show_hidden: false,
+            query: false,
+            copy: false,
+            report_metadata: false,
+            no_metadata: false,
+            verbose: false,
+            handlers: false,
+            handlers_type: WString::new(),
+            description: WString::new(),
+        }
+    }
+}
+
+const NO_METADATA_SHORT: char = 2 as char;
```

Comment:
Oh I quite misunderstood what NONOPTION_CHAR_CODE meant, then?

So it's used to be the char code for the one option that is long-only?

And if we have multiple options that are long-only?

Tbh I prefer naming the constant after the option.

Reply:
It appears I might be one misunderstanding how `NONOPTION_CHAR_CODE` is supposed to be used then, I just assumed it should be used instead of `1 as char / '\x01' / '\u{1}'`, which is how I have used it in `builtins/string/{escape,unescape}` and `builtins/path`.

Looking at the [commit](https://github.com/fish-shell/fish-shell/commit/ecfabf4db860ffd7466fd3c75b22e493638b27f1) that introduced it, it appears `'\x01'` has a special meaning as a non-option argument (duh), but only if the ordering is `REQUIRE_ORDER | RETURN_IN_ORDER`, which is set by `+` and `-` in short_opts respectively. 

When the ordering is the default aka `PERMUTE`, `\x01` does not have that special meaning.

My takeaway here is:

1. I dislike `wgetoptr_t`, since it mixes special meanings with legitimate return values, only configured through a char in a string. A minimal way of improving it would be to instead return an enum like:

```rust
enum WGetOptRetVal {
    LongOpt(usize), // where the usize is the "id" of the long-only option
    Opt(char),
    NonOption,
}
```

We could also make `wopt` accept using a literal number, making it result in a different enum-variant, to make this difference in usage much clearer. (This requires a `IntoOption` trait)

2. We should probably avoid using `\x01` (regardless of whether it is as `NONOPTION_CHAR_CODE` or another named constant) as a short-option for long-only arguments to avoid accidentally mixing this usage, in that way your change to 2 instead of 1 as was present in the original code is good.

Reply:
> I dislike wgetoptr_t, since it mixes special meanings with legitimate return values, only configured through a char in a string

Yup, it has some real crusty C heritage.

>A minimal way of improving it would be to instead return an enum like:

This would return a different value for `-D` and `--details`, which is awkward. I don't *think* you could match like

```rust
LongOpt(id) | Opt(char) => {
}
```

even if you called the variables the same?

> We should probably avoid using \x01 (regardless of whether it is as NONOPTION_CHAR_CODE or another named constant) as a short-option for long-only arguments to avoid accidentally mixing this usage

Agreed. Tho note that "PERMUTE" is the overwhelming majority of uses and changing from that to a different ordering is a pretty big change that is never going to happen for most uses.

Reply:
> This would return a different value for -D and --details, which is awkward. I don't think you could match like
> `LongOpt(id) | Opt(char)`
> even if you called the variables the same?

Ah, I named it a bit unspecificly, I intended `LongOpt` to be `LongOnlyOpt`, if that makes more sense? And then keep the same behaviour for `Opt` if the option has both long and short variants

But a move towards something like your interpretation would make it trivial to determine whether we got the argument as a short- or long-opt, just with another enum-variant. 

We could also add more variants to move the `optarg` to this enum, avoiding all the `w.optarg.unwrap()`-s.